### PR TITLE
Update AdjacentPairs implementations and tests

### PR DESCRIPTION
**Pull request by natecook1000**
Update the `AdjacentPairs` implementation based on code review in #119.

### Checklist
- [x] I've added at least one test that validates that my change is working, if appropriate
- [x] I've followed the code style of the rest of the project
- [x] I've read the [Contribution Guidelines](../blob/main/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary

### CODE REVIEW

File: Sources/Algorithms/AdjacentPairs.swift [hunk 1]
```diff
@@ -214,38 +211,74 @@ extension AdjacentPairsCollection: Collection {
...
+    assert(limit < i)
+        
+    let offset = i == endIndex ? 0 : 1
+    guard let newSecond = base.index(
+      i.first,
+      offsetBy: -(distance - offset),
+      limitedBy: limit.second)
+    else { return nil }
+    let newFirst = base.index(newSecond, offsetBy: -1)
+    precondition(newSecond >= base.startIndex, "Can't move before startIndex")
```
Comment (timvermeulen) [inline Sources/Algorithms/AdjacentPairs.swift:271]:
```suggestion
    precondition(newFirst >= base.startIndex, "Can't move before startIndex")
```

File: Tests/SwiftAlgorithmsTests/AdjacentPairsTests.swift [hunk 2]
```diff
@@ -13,86 +13,68 @@ import XCTest
...
-    let pairs = (1...5).adjacentPairs()
-    let expected = [(4, 5), (3, 4), (2, 3), (1, 2)]
-    var index = pairs.endIndex
-    for iteration in expected.indices {
-      pairs.formIndex(before: &index)
-      XCTAssert(pairs[index] == expected[iteration])
-    }
-    XCTAssertEqual(index, pairs.startIndex)
+  func testIndexTraversals() {
+    validateIndexTraversals((1...5).adjacentPairs())
```
Comment (timvermeulen) [inline Tests/SwiftAlgorithmsTests/AdjacentPairsTests.swift:73]:
```suggestion
    validateIndexTraversals(
      (0..<0).adjacentPairs(),
      (0..<1).adjacentPairs(),
      (0..<2).adjacentPairs(),
      (0..<5).adjacentPairs())
```

File: Sources/Algorithms/AdjacentPairs.swift (hunk 1 above)
Comment (natecook1000) [reply Sources/Algorithms/AdjacentPairs.swift:271]: 👍🏻 We really need `expectCrashLater`.... I manually exercised some of these preconditions but didn’t hit this one!

**Comment by natecook1000** [comment]
@swift-ci Please test

**Pull request merged**